toolchain/semantics/semantics_handle.cpp

```diff
@@ -492,6 +526,46 @@ auto SemanticsHandleSelfValueIdentifier(SemanticsContext& context,
   return context.TODO(parse_node, "HandleSelfValueIdentifier");
 }
 
+auto SemanticsHandleShortCircuitOperand(SemanticsContext& context,
+                                        ParseTree::Node parse_node) -> bool {
+  // Convert the condition to `bool`.
+  auto cond_value_id = context.node_stack().Pop<SemanticsNodeId>();
+  cond_value_id = context.ImplicitAsBool(parse_node, cond_value_id);
+
+  // Compute the branch value: the condition for `and`, inverted for `or`.
+  auto token = context.parse_tree().node_token(parse_node);
+  SemanticsNodeId branch_value_id = SemanticsNodeId::Invalid;
+  switch (auto token_kind = context.tokens().GetKind(token)) {
+    case TokenKind::And:
+      branch_value_id = cond_value_id;
+      break;
+
+    case TokenKind::Or:
+      branch_value_id = context.AddNode(SemanticsNode::UnaryOperatorNot::Make(
+          parse_node, context.semantics_ir().GetNode(cond_value_id).type_id(),
+          cond_value_id));
```

Comment:
When you have a moment, I'd like to discuss whether we should be hewing closer to `br` and `phi` rather than the current `BranchIf` + `BranchWithArg` + `BlockArg` approach. Seeing this code made me wonder a bit why the output paths aren't just swapped.

I consider this non-blocking as it'd change existing merged code too, and perhaps it was already considered.

Reply:
I discussed this at some length with @chandlerc in the context of the `if` expressions PR, and he proposed we use block arguments rather than `phi` nodes. Definitely open to reconsidering it, but I'd want @chandlerc involved in the disucssion, and I think we shouldn't block on it for now.

We could swap over the branch destinations instead of generating a `not`, but that would require something like a `BranchIfWithArg` (which awkwardly needs three operands and therefore needs separate storage) or generation of an additional block. I think those are reasonable options to consider, but maybe it'd make some sense to wait until we see what we need from other branching constructs?

Reply:
I'm considering this resolved for now; probably see how lowering goes and also discuss with chandlerc.